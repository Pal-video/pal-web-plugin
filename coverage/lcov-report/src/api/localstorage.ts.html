
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/api/localstorage.ts</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> / <a href="index.html">src/api</a> localstorage.ts</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">64.28% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>9/14</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">55.55% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>5/9</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">50% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>2/4</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">64.28% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>9/14</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line medium'></div>
    <pre><table class="coverage">
<tr><td class="line-count quiet"><a name='L1'></a><a href='#L1'>1</a>
<a name='L2'></a><a href='#L2'>2</a>
<a name='L3'></a><a href='#L3'>3</a>
<a name='L4'></a><a href='#L4'>4</a>
<a name='L5'></a><a href='#L5'>5</a>
<a name='L6'></a><a href='#L6'>6</a>
<a name='L7'></a><a href='#L7'>7</a>
<a name='L8'></a><a href='#L8'>8</a>
<a name='L9'></a><a href='#L9'>9</a>
<a name='L10'></a><a href='#L10'>10</a>
<a name='L11'></a><a href='#L11'>11</a>
<a name='L12'></a><a href='#L12'>12</a>
<a name='L13'></a><a href='#L13'>13</a>
<a name='L14'></a><a href='#L14'>14</a>
<a name='L15'></a><a href='#L15'>15</a>
<a name='L16'></a><a href='#L16'>16</a>
<a name='L17'></a><a href='#L17'>17</a>
<a name='L18'></a><a href='#L18'>18</a>
<a name='L19'></a><a href='#L19'>19</a>
<a name='L20'></a><a href='#L20'>20</a>
<a name='L21'></a><a href='#L21'>21</a>
<a name='L22'></a><a href='#L22'>22</a>
<a name='L23'></a><a href='#L23'>23</a>
<a name='L24'></a><a href='#L24'>24</a>
<a name='L25'></a><a href='#L25'>25</a>
<a name='L26'></a><a href='#L26'>26</a>
<a name='L27'></a><a href='#L27'>27</a>
<a name='L28'></a><a href='#L28'>28</a>
<a name='L29'></a><a href='#L29'>29</a>
<a name='L30'></a><a href='#L30'>30</a>
<a name='L31'></a><a href='#L31'>31</a>
<a name='L32'></a><a href='#L32'>32</a>
<a name='L33'></a><a href='#L33'>33</a>
<a name='L34'></a><a href='#L34'>34</a>
<a name='L35'></a><a href='#L35'>35</a>
<a name='L36'></a><a href='#L36'>36</a>
<a name='L37'></a><a href='#L37'>37</a>
<a name='L38'></a><a href='#L38'>38</a>
<a name='L39'></a><a href='#L39'>39</a>
<a name='L40'></a><a href='#L40'>40</a>
<a name='L41'></a><a href='#L41'>41</a>
<a name='L42'></a><a href='#L42'>42</a>
<a name='L43'></a><a href='#L43'>43</a>
<a name='L44'></a><a href='#L44'>44</a>
<a name='L45'></a><a href='#L45'>45</a>
<a name='L46'></a><a href='#L46'>46</a>
<a name='L47'></a><a href='#L47'>47</a>
<a name='L48'></a><a href='#L48'>48</a>
<a name='L49'></a><a href='#L49'>49</a>
<a name='L50'></a><a href='#L50'>50</a></td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import {PalSession} from './models/session.model';
&nbsp;
const sessionKey = 'PalSession';
&nbsp;
export class LocalstorageService {
  private _session: PalSession | null = null;
&nbsp;
  constructor() {
    const saveSession =
      sessionStorage.getItem(sessionKey) || localStorage.getItem(sessionKey);
    <span class="missing-if-branch" title="if path not taken" >I</span>if (saveSession) {
<span class="cstat-no" title="statement not covered" >      this._session = JSON.parse(saveSession);</span>
    }
  }
&nbsp;
  /**
   * Check if user has a session stored
   * @returns True if the user session is set.
   */
<span class="fstat-no" title="function not covered" >  hasSession(</span>): boolean {
<span class="cstat-no" title="statement not covered" >    return !!this.session;</span>
  }
&nbsp;
  /**
   * Gets the user session stored.
   * @return The user session or null if no session found.
   */
<span class="fstat-no" title="function not covered" >  get s</span>ession(): PalSession | null {
<span class="cstat-no" title="statement not covered" >    return this._session;</span>
  }
&nbsp;
  /**
   * Sets the user session.
   * Otherwise, the credentials are only persisted for the current session.
   * @param session The user session.
   * @param remember True to remember credentials across sessions (default true).
   */
  setSession(session?: PalSession, remember?: boolean) {
    this._session = session || <span class="branch-1 cbranch-no" title="branch not covered" >null;</span>
&nbsp;
    if (session) {
      const storage = remember ? <span class="branch-0 cbranch-no" title="branch not covered" >localStorage </span>: sessionStorage;
      storage.setItem(sessionKey, JSON.stringify(session));
    } else <span class="missing-if-branch" title="else path not taken" >E</span>{
<span class="cstat-no" title="statement not covered" >      sessionStorage.removeItem(sessionKey);</span>
<span class="cstat-no" title="statement not covered" >      localStorage.removeItem(sessionKey);</span>
    }
  }
}
&nbsp;</pre></td></tr></table></pre>

                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at Wed Jun 15 2022 11:20:10 GMT+0200 (heure d’été d’Europe centrale)
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    