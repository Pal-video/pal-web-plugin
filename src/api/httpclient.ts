import * as rm from 'typed-rest-client/RestClient';

export class HttpClient {

    private baseUrl: string;

    private apiKey: string;

    private httpClient: rm.RestClient;

    constructor({ baseUrl, apiKey }: { baseUrl: string; apiKey: string }) {
        this.baseUrl = baseUrl;
        this.apiKey = apiKey;
        this.httpClient = new rm.RestClient(
            'pal',
            this.baseUrl,
        );
        let options = {
            maxSockets: 1,
            headers: {
                'Authorization': `Bearer ${this.apiKey}`,
            }
        };
        this.httpClient.client.requestOptions = options;
    }

    public get<T>(method: string): Promise<T | null> {
        return this.httpClient.get<T>(method).then(res => res.result);
    }

    public post<T>(method: string, data: unknown): Promise<T | null> {
        return this.httpClient
            .create<T>(method, data)
            .then(res => {
                if (res.result && res.statusCode < 200 && res.statusCode >= 300) {
                    throw new Error(`
                        Pal returned an error:
                        url: ${this.baseUrl}${method}
                        reason: ${res.statusCode}
                    `);
                }
                return res.result;
            });
    }

    public put<T>(method: string, data: unknown): Promise<T | null> {
        return this.httpClient
            .update<T>(method, data)
            .then(res => {
                if (res.result && res.statusCode < 200 && res.statusCode >= 300) {
                    throw new Error(`
                        Pal returned an error:
                        url: ${this.baseUrl}${method}
                        reason: ${res.statusCode}
                    `);
                }
                return res.result;
            });

    }
}
