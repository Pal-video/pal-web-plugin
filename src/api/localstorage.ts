import {PalSession} from './models/session.model';

const sessionKey = 'PalSession';

export class LocalstorageService {
  private _session: PalSession | null = null;

  constructor() {
    const saveSession =
      sessionStorage.getItem(sessionKey) || localStorage.getItem(sessionKey);
    if (saveSession) {
      this._session = JSON.parse(saveSession);
    }
  }

  /**
   * Check if user has a session stored
   * @returns True if the user session is set.
   */
  hasSession(): boolean {
    return !!this.session;
  }

  /**
   * Gets the user session stored.
   * @return The user session or null if no session found.
   */
  get session(): PalSession | null {
    return this._session;
  }

  /**
   * Sets the user session.
   * Otherwise, the credentials are only persisted for the current session.
   * @param session The user session.
   * @param remember True to remember credentials across sessions (default true).
   */
  setSession(session?: PalSession, remember?: boolean) {
    this._session = session || null;

    if (session) {
      const storage = remember ? localStorage : sessionStorage;
      storage.setItem(sessionKey, JSON.stringify(session));
    } else {
      sessionStorage.removeItem(sessionKey);
      localStorage.removeItem(sessionKey);
    }
  }
}
