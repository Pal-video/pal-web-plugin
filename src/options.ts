import {PlatformTypes} from './api/models/session.model';

export interface PalOptions {
  apiKey: string;
  serverUrl: string;
  platform?: PlatformTypes;
}
