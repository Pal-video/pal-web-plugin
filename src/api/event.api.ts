import { HttpClient } from "./httpclient";
import { PalEvent, PalEventContext, PalEventType } from "./models/event.model";
import { PalSession } from "./models/session.model";
import { PalVideoTrigger } from "./models/video_trigger.model";

export class PalEventApi {

    constructor(
        private httpClient: HttpClient
    ) { }

    async logCurrentScreen(session: PalSession, name: string): Promise<PalVideoTrigger | null> {
        return this._logEvent(session, <PalEvent>{
            name: name,
            type: PalEventType.setScreen,
        });
    }

    private async _logEvent(session: PalSession, event: PalEvent): Promise<PalVideoTrigger | null> {
        const eventContext = <PalEventContext>{
            sessionUId: session.uid,
            ...event
        };
        const response = await this.httpClient.post<PalVideoTrigger>('/eventlogs', eventContext);
        if (response && response.videoId) {
            return response;
        }
        return null;
    }

}