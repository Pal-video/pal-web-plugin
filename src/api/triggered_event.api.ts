import { HttpClient } from "./httpclient";
import { PalSession } from "./models/session.model";
import { ChoiceItem } from "./models/survey.model";
import { VideoTriggerEvent, VideoTriggerEvents } from "./models/triggered_event.model";
import { PalVideoTrigger } from "./models/video_trigger.model";

export class PalTriggeredEventApi {
    constructor(
        private httpClient: HttpClient
    ) { }

    async save(
        triggeredVideoEvent: VideoTriggerEvent,
        eventlogId?: string | null,
    ): Promise<void> {
        if (!eventlogId) {
            return;
        }
        await this.httpClient.post(`/eventlogs/${eventlogId}`, triggeredVideoEvent);
    }

    async saveSurveyAnswer(
        session: PalSession,
        triggeredVideo: PalVideoTrigger,
        choice: ChoiceItem
    ): Promise<void> {
        const event = <VideoTriggerEvent>{
            type: VideoTriggerEvents.answer,
            time: new Date(), //FIXME use dayJS
            sessionId: session.uid,
            args: {
                "answer": choice.code
            }
        };
        await this.save(event, triggeredVideo.eventLogId);
    }

}