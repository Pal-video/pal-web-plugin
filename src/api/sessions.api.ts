import {HttpClient} from './httpclient';
import {LocalstorageService} from './localstorage';
import {PalSession, PalSessionRequest} from './models/session.model';

export class SessionsApi {
  constructor(
    private localstorageService: LocalstorageService,
    private httpClient: HttpClient
  ) {}

  async createSession(request: PalSessionRequest): Promise<PalSession> {
    const response = await this.httpClient.post<PalSession>('/sessions', {
      frameworkType: request.frameworkType,
      platform: request.platform,
    });
    this.localstorageService.setSession(response);
    return response;
  }

  async getSession(): Promise<PalSession | null> {
    return this.localstorageService.session;
  }

  clearSession() {
    return this.localstorageService.clear();
  }
}
