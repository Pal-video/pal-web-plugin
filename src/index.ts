import { PalEventApi } from './api/event.api';
import { HttpClient } from './api/httpclient';
import { LocalstorageService } from './api/localstorage';
import {
    PalSession,
    PalSessionRequest,
    PlatformTypes,
} from './api/models/session.model';
import { ChoiceItem } from './api/models/survey.model';
import { VideoTriggerEvent, VideoTriggerEvents } from './api/models/triggered_event.model';
import { PalVideoTrigger } from './api/models/video_trigger.model';
import { SessionsApi } from './api/sessions.api';
import { PalTriggeredEventApi } from './api/triggered_event.api';
import { PalOptions, PRODUCTION_SERVER_URL } from './options';
import { PalSdk, ShowSurveyParams, ShowVideoOnlyParams } from './sdk/palsdk';

export class Pal {

    private palSdk: PalSdk;

    private options!: PalOptions;

    private hasInitialized = false;

    private static _instance: Pal | null;

    static getInstance(): Pal {
        if (!this._instance) {
            throw new Error(
                'Pal has not been initialized, please call createInstance'
            );
        }
        return this._instance;
    }

    static createInstance(options: PalOptions): Pal {
        const httpClient = new HttpClient({
            baseUrl: options.serverUrl ?? PRODUCTION_SERVER_URL,
            apiKey: options.apiKey,
        });
        const sessionsApi = new SessionsApi(new LocalstorageService(), httpClient);
        const eventsApi = new PalEventApi(httpClient);
        const triggeredEventApi = new PalTriggeredEventApi(httpClient);
        this._instance = new Pal(
            sessionsApi,
            eventsApi,
            options,
            new PalSdk(),
            triggeredEventApi
        );
        return this._instance;
    }

    constructor(
        private sessionsApi: SessionsApi,
        private eventsApi: PalEventApi,
        options: PalOptions,
        palSdk: PalSdk,
        private triggeredEventApi?: PalTriggeredEventApi,
    ) {
        this.options = options;
        this.palSdk = palSdk;
    }

    async initialize(): Promise<void> {
        if (this.hasInitialized) {
            return;
        }
        try {
            const hasSessionInStorage = (await this.sessionsApi.getSession()) != null;
            if (hasSessionInStorage) {
                this.hasInitialized = true;
                return;
            }
            await this.sessionsApi.createSession(<PalSessionRequest>{
                frameworkType: 'JAVASCRIPT',
                platform: this.options.platform ?? PlatformTypes.web,
            });
            this.hasInitialized = true;
        } catch (err) {
            console.error("cannot initialize Pal plugin", err);
        }
    }

    async getSession(): Promise<PalSession | null> {
        if (!this.hasInitialized) {
            throw new Error('Pal has not been initialized');
        }
        return this.sessionsApi.getSession();
    }

    async clearSession(): Promise<void> {
        if (!this.hasInitialized) {
            throw new Error('Pal has not been initialized');
        }
        this.sessionsApi.clearSession();
    }

    /**
     * send the screen view event to the server
     * if a video is triggered, it will be returned
     * depending on configuration
     * - user will see video on first time this screen as been seen
     * Or each time this screen is visited
     * 
     * each time we log a new screen we remove old videos from the page
     */
    async logCurrentScreen(name: string) {
        this.palSdk.clear();

        const session = await this.getSession();
        if (!session) {
            throw new Error('Pal has not been initialized');
        }
        try {
            const triggeredVideo = await this.eventsApi.logCurrentScreen(session, name);
            if (!triggeredVideo) {
                return;
            }
            let isSurveyType = triggeredVideo.survey != null;
            let isTalkType = triggeredVideo.survey == null;
            // console.log('triggeredVideo', triggeredVideo.videoUrl);
            if (isTalkType) {
                this.palSdk.showVideoOnly(
                    <ShowVideoOnlyParams>{
                        videoUrl: triggeredVideo.videoUrl,
                        minVideoUrl: triggeredVideo.videoThumbUrl,
                        userName: triggeredVideo.videoSpeakerName,
                        companyTitle: triggeredVideo.videoSpeakerRole,
                        onExpand: () => this._sendTriggeredEvent(session, triggeredVideo, VideoTriggerEvents.minVideoOpen),
                        onSkip: () => this._sendTriggeredEvent(session, triggeredVideo, VideoTriggerEvents.videoSkip),
                        onVideoEnd: () => this._sendTriggeredEvent(session, triggeredVideo, VideoTriggerEvents.videoViewed),
                    }
                );
            } else if (isSurveyType) {
                const options: ChoiceItem[] = [];
                const keys = Object.keys(triggeredVideo.survey!.options);
                for (let i = 0; i < keys.length; i++) {
                    options.push(<ChoiceItem>{
                        code: keys[i],
                        text: triggeredVideo.survey!.options[keys[i]],
                    });
                }
                this.palSdk.showSingleChoiceSurvey(
                    <ShowSurveyParams>{
                        videoUrl: triggeredVideo.videoUrl,
                        minVideoUrl: triggeredVideo.videoThumbUrl,
                        userName: triggeredVideo.videoSpeakerName,
                        companyTitle: triggeredVideo.videoSpeakerRole,
                        question: triggeredVideo.survey!.question,
                        choices: options,
                        onTapChoice: (choice) => this.triggeredEventApi?.saveSurveyAnswer(session, triggeredVideo, choice),
                        onExpand: () => this._sendTriggeredEvent(session, triggeredVideo, VideoTriggerEvents.minVideoOpen),
                        onSkip: () => this._sendTriggeredEvent(session, triggeredVideo, VideoTriggerEvents.videoSkip),
                        onVideoEnd: () => this._sendTriggeredEvent(session, triggeredVideo, VideoTriggerEvents.videoViewed),
                    }
                );
            }
        } catch (err) {
            console.error("logCurrentScreen");
            console.error(err);
        }
    }

    private async _sendTriggeredEvent(
        session: PalSession,
        triggeredVideo: PalVideoTrigger,
        eventType: VideoTriggerEvents,
    ) {
        this.triggeredEventApi?.save(
            <VideoTriggerEvent>{
                type: eventType,
                time: new Date(), //FIXME use dayJS
                sessionId: session.uid,
            },
            triggeredVideo.eventLogId
        );
    }

}
