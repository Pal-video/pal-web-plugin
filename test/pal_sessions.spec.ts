import 'jest';
import Pal from '../src';
import {HttpClient} from '../src/api/httpclient';
import {LocalstorageService} from '../src/api/localstorage';
import {PalSession, PlatformTypes} from '../src/api/models/session.model';
import {SessionsApi} from '../src/api/sessions.api';
import {PalOptions} from '../src/options';

describe('initialize Pal with session', () => {
  let pal: Pal;

  let httpClient: HttpClient;

  let sessionsApi: SessionsApi;

  beforeEach(() => {
    const options = <PalOptions>{
      apiKey: '',
      serverUrl: '',
      platform: PlatformTypes.web,
    };
    httpClient = new HttpClient({baseUrl: '', apiKey: ''});
    sessionsApi = new SessionsApi(new LocalstorageService(), httpClient);
    pal = new Pal(sessionsApi, options);
  });

  beforeEach(() => {
    localStorage.clear();
    sessionStorage.clear();
    jest.resetAllMocks();
  });

  test('creates a session', async () => {
    const createSessionSpy = jest.spyOn(sessionsApi, 'createSession');
    jest
      .spyOn(httpClient, 'post')
      .mockResolvedValue(Promise.resolve(<PalSession>{uid: '8093283093'}));

    // init session
    await pal.initialize();

    // session api has be called and session has been stored
    expect(createSessionSpy).toHaveBeenCalled();
    expect(sessionsApi.getSession()).resolves.toEqual(<PalSession>{
      uid: '8093283093',
    });
  });

  test('get a session without initialize should throw', async () => {
    const createSessionSpy = jest.spyOn(sessionsApi, 'createSession');

    expect(createSessionSpy).toBeCalledTimes(0);
    expect(pal.getSession()).rejects.toThrowErrorMatchingSnapshot(
      'Pal has not been initialized'
    );
  });

  test('creates a session, call clear session => get session returns null', async () => {
    const createSessionSpy = jest.spyOn(sessionsApi, 'createSession');
    jest
      .spyOn(httpClient, 'post')
      .mockResolvedValue(Promise.resolve(<PalSession>{uid: '8093283093'}));

    // init session and clear session
    await pal.initialize();
    await pal.clearSession();

    // session api has be called and session has been stored
    expect(sessionsApi.getSession()).resolves.toBeNull();
  });
});
