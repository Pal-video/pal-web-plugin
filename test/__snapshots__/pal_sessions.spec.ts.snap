// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initialize Pal plugin get a session without initialize should throw: Pal has not been initialized 1`] = `"Pal has not been initialized"`;

exports[`initialize Pal with session get a session without initialize should throw: Pal has not been initialized 1`] = `"Pal has not been initialized"`;
